### ISSUE TITLE
Overriding getters and setters for particular fields

### ISSUE THREAD

**Issue by serkandurusoy** [jagi/meteor-astronomy#176]
A schema is a living thing and often sees evolutions through which some of the fields change in order to adapt to new business requirements.

In such cases, I would expect a full-featured model layer to provide me means to override getters and setters of existing fields.

For example a previously existing field may get swapped out for another one and instead of migrating old data and/or refactoring the complete application where there are references to that field, I might want to opt for changing its getter and setter to for example check for the createdAt property and feed its return function the new field's value for newer documents.

There are so many use cases for this. It basically truly decouples the application and the data not just as a model layer, but also as a data access and persistence layer.

Is there currently a way we can achieve this, or do you plan on providing such a feature?

Thanks!

**Comment by jagi** [comment]
Yes I was thinking about this feature but haven't come with a good solution so far. One thing is having a different field name in a model and different field name in a database:

```js
User = Astro.Class({
  /* ... */
  fields: {
    lastName: {
      type: 'string',
      map: 'surname'
    }
  }
});
```

What it could do, is storing a value of the `lastName` field in the `surname` field in a collection. I don't know if I would use the `map` property. Maybe there is a better name for this option.

It would solve a problem you are talking about. However, right now it's not possible. I was thinking about introducing it in Astronomy 2.0 but if there is a demand for that maybe I could introduce it in the current version.

**Comment by serkandurusoy** [comment]
Hmm, this would be a clever implementation.

But why not just provide exclusie `get` and `set` properties that are optional, but if present, applied before passing onto the database, possibly also through the `map`

and `map` itself could accept a function that returns a string. This way, data persistence endpoints could be changed on the fly!

**Comment by jagi** [comment]
I prefer the `map` property that can take a function that returns a string which will be the field name in a database. I'm going to implement it in near future.

**Comment by awei01** [comment]
just to throw another idea out there, how about a "migrations" behavior that will allow the developer to define different versions of the schema. Along with each schema version is a set of transformations that will run on fetching the document. Once the doc is stored, the behavior will stamp the document with a `migrationVersion` field. I think this would help extract the migration logic out of the `fields` declaration. thoughts?

**Comment by serkandurusoy** [comment]
Well it could be handy for scenarios where multiple versions coexist and migrations get done incrementally as data gets piped.

**Comment by jagi** [comment]
If it goes about migration I was planning this feature from the early beginning. It will definitely appear but I don't know in what form. Let me give you some background on it. There are two possible ways of dealing with migrations.

**1. Do not modify database**

This approach can be familiar to Android developers who were playing with Google Endpoints. You can define several versions of the model. Each application version can have a different model. Let's say we have an application v1.0 that has a model:

```js
User
- firstName
- lastName
- birthDate
```

Every document stored in a database has such form. No in v2.0 we introduce a new field `address` which is mandatory.

```js
User
- firstName
- lastName
- birthDate
- address
```

In 1.0 version of app this field will be just ignored, but what about 2.0? Some document stored in a database don't have this field. That's why a model for v2.0 should fill this field with some value that will not break the app.

What are pros and cons of such approach?

**Pros:** You can have several application versions that are working even that they have a different schema. Many users of Android have the automatic application update turned off, so they can still work on version 1.0 when there is 2.0 on the market.

**Cons:** Mess in a database. This way, we have tons of documents that have different fields.

**2. Modify database**

IMHO this approach is more well fitted to use in Meteor. We can define several models, but when we run migration, the database gets updated. If we take an example from the above, it would add the `address` field to each document in a database. Of course it would still be hard to set value for this field, but we avoid a mess in a database. Why is it better in Meteor? Because Meteor application requires Internet connection, so there no situation like "I have the old model in my app version". I don't know how situation look like if we consider Android application and Cordova usage, but let me know.

**Pros:** No mess in a database. Faster documents fetching. Without an another migration layer between database and model, we don't have to execute extra calculations. Fetching documents from the collection is the most common action, so it's very important to make it run fast.

**Comment by serkandurusoy** [comment]
Well both methods have their usecases, in fact, a hybrid, incremental migration approach should be listed as a third possible option.

I don't particularly agree that method 2 is better suited with meteor apps. The connectedness of an app does play a role in its migration strategy, but so does the database size.

Imagine you have an app with billions of records, something like Kadira, and you want to introduce a new feature that requires a new model. But obviously, you won't be able to migrate old records, so the app should play nice with both the old model and the new one at the same time.

Or imagine a scenario where data gets classified as live/recent and outdated. Outdated data may never get called, but if it does (imagine the facebook timeline entries from five years ago, who scrolls that far? but someone can!) it should be migrated on the fly. So the app needs to support both the old and the new model and whenever data that belongs to the old model, it gets migrated to the new model on the fly, on first access. You perhaps can't do it for the complete data since it is very large, but  ad-hoc migration can be a good option.

Anyway, if we are speaking a truly flexible migration system, we should not be opinionated. But it is your package, so you are entitled to an opinion :)))

**Comment by serkandurusoy** [comment]
@awei01 that's correct, but in all honesty, I presented this example to be trivial and representative of what an override would achieve. 

Furthermore, the *event* subsystem is more efficient for ad-hoc transformations wheres overrides represent the *current* state of an entity. Besides, there's also the setters, so the need for some symmetry.

**Comment by awei01** [comment]
@serkandurusoy got it.

**Comment by jagi** [comment]
Ok let me say a little bit more about Astronomy 2.0. I'm removing all of the following methods: `set()`, `get()`, `push()`, `pull()`, `inc()`. So there will be not `beforeSet` and `afterSet` events. You will be able to manipulate data directly.

```js
doc.field = 'test';
doc.array.push('test');
doc.array = doc.array.slice(1, 2);
```

So you can use standard way of modifying data and every possible library like underscore or lodash. As I said before I'm going to introduce the `map` property in a field definition.

```js
fields: {
  firstName: {
    type: 'string',
    map: 'name' // The real name of the field in a database
  }
}
```

or 

```js
fields: {
  firstName: {
    type: 'string',
    map: function() { return 'name'; } // Computed field name
  }
}
```

It will solve the problem of having different fields name in model and in database. If it goes about lack of some field or extra fields. There are always `beforeInit` and `afterInit` events which can deal with it.

I don't know if there is a need for a migration system when having all the mentioned above features. If Meteor with Cordova works as been said, then there is not need for storing many different versions of the class.

**Comment by serkandurusoy** [comment]
sounds nice. could you possibly at least split map into mapto/mapfrom so
that one is used for reading and the other for writing?

Sent from my iPhone



On Nov 19, 2015, at 10:53, jagi <notifications@github.com> wrote:

Ok let me say a little bit more about Astronomy 2.0. I'm removing all of
the following methods: set(), get(), push(), pull(), inc(). So there will
be not beforeSet and afterSet events. You will be able to manipulate data
directly.

doc.field = 'test';
doc.array.push('test');
doc.array = doc.array.slice(1, 2);

So you can use standard way of modifying data and every possible library
like underscore or lodash. As I said before I'm going to introduce the map
property in a field definition.

fields: {
  firstName: {
    type: 'string',
    map: 'name' // The real name of the field in a database
  }
}

or

fields: {
  firstName: {
    type: 'string',
    map: function() { return 'name'; } // Computed field name
  }
}

It will solve the problem of having different fields name in model and in
database. If it goes about lack of some field or extra fields. There are
always beforeInit and afterInit events which can deal with it.

I don't know if there is a need for a migration system when having all the
mentioned above features. If Meteor with Cordova works as been said, then
there is not need for storing many different versions of the class.

—
Reply to this email directly or view it on GitHub
<https://github.com/jagi/meteor-astronomy/issues/176#issuecomment-157992667>
.

**Comment by jagi** [comment]
I think it does not matter if you are reading or writing. It's like the name of a field in an app is the one defined in the schema. The name of a field under which it's stored in a collection is the one from the map property.

**Comment by serkandurusoy** [comment]
consider a migration scenario where I need to read from the new field if
exists or the old field if not but always write to the new field.

Sent from my iPhone



On Nov 19, 2015, at 16:26, jagi <notifications@github.com> wrote:

I think it does not matter if you are reading or writing. It's like the
name of a field in an app is the one defined in the schema. The name of a
field under which it's stored in a collection is the one from the map
property.

—
Reply to this email directly or view it on GitHub
<https://github.com/jagi/meteor-astronomy/issues/176#issuecomment-158072306>
.

**Comment by jagi** [comment]
You will be able to achieve that by using `beforeInit` event.

```jss
events: {
  beforeInit: function(e) {
    var attrs = e.data.attrs;
    if (!_.has(attrs, 'newField') && _.has(attrs, 'oldField') {
      attrs.newField = attrs.oldField;
    }
  }
}
```

Or something similar. Maybe it's not the most clean solution but would work.

The scenario you are talking about is not solvable by introducing mapTo/mapFrom properties. Or am I missing something?

**Comment by awei01** [comment]
Perhaps a `resolve()` method in the definition that gets passed the actual collection data?

```
//version 1

fields: {
  foo: "string",  
}

// version 2
fields: {
    newFoo: {
         type: "number",
         resolve: function(data) {
             if (_.has(data, 'foo')) {
                  return Number(data.foo);
             }
             return null;
         }
    }
```

**Comment by jagi** [comment]
Hmm I had to write it down to get the idea better:

```js
// DATA IN COLLECTION
[{ // old document
  name: 'a',
  oldField: 'oldValue',
}, { // new document
  name: 'b',
  newField: 'newValue'
}]

// OLD SCHEMA
fields: {
  'name': 'string',
  'oldField': {
    type: 'string'
  }
}

// NEW SCHEMA
fields: {
  'name': 'string',
  'newField': {
    type: 'string',
    map: 'oldField'
  }
}

Collection.findOne({name: 'a'}); // { name: 'a', newField: 'oldValue' }
Collection.findOne({name: 'b'}); // { name: 'b', newField: 'newValue' }
```

and after thinking about it, in deed instead of `map`, `mapTo/mapFrom` properties there should be some kind of function (like `resolve`) that looks for the value in the data coming from the collection and I think it should be something sitting between collection and schema/fields.

```js
Astro.Class({
  name: 'Name',
  collection: Collection,
  fields: {
    newField: 'string',
    name: 'string
  },
  mapping: function(attrs) {
    if (!_.has(attrs, 'newField') && _.has(attrs, 'oldField')) {
      attrs.newField = attrs.oldField;
    }
    return attrs;
  }
});
```

**Comment by talha-asad** [comment]
I like the resolve approach. Will this mapping function run on ```init```?

**Comment by jagi** [comment]
The mapping function will be run before initialisation. So in the `init` you will have already mapped document.

**Comment by jagi** [comment]
The `resolve` function is available in the field definition in Astronomy 2.0. So I'm closing this issue for now.
